### PR TITLE
NEW Send an email notification when a job is broken

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/symbiote/silverstripe-queuedjobs/issues/435

There is an email sent for 'stalled' jobs, however there is not an email sent for 'broken' jobs. There is existing logic to ensure that there is only a single notification sent, however no email is actually sent, instead the only notification goes to the logger. An [inline comment](https://github.com/symbiote/silverstripe-queuedjobs/blob/4/src/Services/QueuedJobService.php#L500) says "email" so looks like the intention all along was to send an email.

Testing steps

Create a vanilla install of silverstripe/installer + this PR

**_config/myemail.yml**
```yml
---
Name: myemail
After:
  - '#project-emailconfig'
---
SilverStripe\Control\Email\Email:
  admin_email:
    admin-email@example.com: 'Admin-email'
# + something to setup mailhog / mailcatcher / etc
```

**MyBrokenJob.php**
```php
<?php

use Symbiote\QueuedJobs\Services\AbstractQueuedJob;
use Symbiote\QueuedJobs\Services\QueuedJob;

class MyBrokenJob extends AbstractQueuedJob
{
    public function getTitle()
    {
        return 'My broken job';
    }

    public function getJobType() {
        return QueuedJob::QUEUED;
    }

    public function process()
    {
        throw new Exception('MY BROKEN JOB EXCEPTION');
    }
}
```

In `admin/queuedjobs` create a new 'My broken job'

In console:
`vendor/bin/sake dev/tasks/ProcessJobQueueTask`
`vendor/bin/sake dev/tasks/CheckJobHealthTask`

There should be an 'Broken job' email that was sent

### CODE REVIEW

File: src/Services/QueuedJobService.php [hunk 1]
```diff
@@ -519,6 +519,25 @@ public function checkJobHealth($queue = null)
             ]
         );
 
+        // Send broken job email notification
+        $subject = _t(__CLASS__ . '.BROKEN_JOB', 'Broken job');
+        $message = sprintf(_t(
```
Comment (emteknetnz) [inline src/Services/QueuedJobService.php]: Using a combination of sprintf() with _t() is pretty standard elsewhere in the codebase.

You can used named variables with _t(), though that's much less common

File: src/Services/EmailService.php [hunk 2]
```diff
@@ -35,11 +34,6 @@ public function __construct()
         );
     }
 
-    /**
-     * @param array $jobConfig
```
Comment (emteknetnz) [inline src/Services/EmailService.php:39]: Unrelated refactoring, methods are strongly typed do not need docblock types

File: src/Services/EmailService.php [hunk 3]
```diff
@@ -6,7 +6,6 @@
 use SilverStripe\Control\Email\Email;
 use SilverStripe\Core\Config\Config;
 use SilverStripe\Core\Injector\Injectable;
-use SilverStripe\Subsites\Model\Subsite;
```
Comment (emteknetnz) [inline src/Services/EmailService.php:9]: Unrelated refactoring, unused import

File: src/Services/QueuedJobService.php [hunk 4]
```diff
@@ -519,6 +519,24 @@ public function checkJobHealth($queue = null)
...
+                'The following job(s) appear to be broken.',
+                'Log in to %s to see further details and take any necessary actions.',
+            ])
+        ), Director::absoluteBaseURL());
+        foreach ($brokenJobs as $brokenJob) {
+            $message .= sprintf("\n- %s (ID: %s)", $brokenJob->JobTitle, $brokenJob->ID);
+        }
+        $email = EmailService::singleton()->createBrokenJobReport($subject, $message);
+        if ($email) {
+            $email->sendPlain();
```
Comment (ChloeSartorelli) [inline src/Services/QueuedJobService.php]: Why specifically `sendPlain()`? The other emails in this file use `send()`.

File: src/Services/QueuedJobService.php [hunk 5]
```diff
@@ -519,6 +519,24 @@ public function checkJobHealth($queue = null)
...
+        $message = sprintf(_t(
+            __CLASS__ . '.BROKEN_JOBS_MSG',
+            implode(' ', [
+                'The following job(s) appear to be broken.',
+                'Log in to %s to see further details and take any necessary actions.',
+            ])
+        ), Director::absoluteBaseURL());
+        foreach ($brokenJobs as $brokenJob) {
+            $message .= sprintf("\n- %s (ID: %s)", $brokenJob->JobTitle, $brokenJob->ID);
+        }
```
Comment (ChloeSartorelli) [inline src/Services/QueuedJobService.php]: Probably better to use an email template, same as the other emails sent from this module.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Works well locally, just a couple of small things to fix.

File: src/Services/QueuedJobService.php (hunk 4 above)
Comment (emteknetnz) [reply src/Services/QueuedJobService.php]: Updated

File: src/Services/QueuedJobService.php (hunk 5 above)
Comment (emteknetnz) [reply src/Services/QueuedJobService.php]: Updated

File: templates/QueuedJobsBrokenJobs.ss [hunk 6]
```diff
@@ -0,0 +1 @@
+$Message.RAW
```
Comment (ChloeSartorelli) [inline templates/QueuedJobsBrokenJobs.ss]: Should be more like the `QueuedJobsStalledJob.ss` file.

```suggestion
Hi,

$Message

<% loop $Jobs $>
- $JobTitle (ID: $ID)
<% end_loop %>

Log in to $Site to see further details and take any necessary actions.
```

File: src/Services/QueuedJobService.php [hunk 7]
```diff
@@ -519,6 +519,26 @@ public function checkJobHealth($queue = null)
...
+            implode(' ', [
+                'The following job(s) appear to be broken.',
+                'Log in to %s to see further details and take any necessary actions.',
+            ])
+        ), Director::absoluteBaseURL()));
+        foreach ($brokenJobs as $brokenJob) {
+            // Using htmlentities() because of the use of $Message.RAW in QueuedJobsBrokenJobs.ss
+            // in order for the <br> tags to render as HTML
+            $message .= sprintf("<br>- %s (ID: %s)", htmlentities($brokenJob->JobTitle), $brokenJob->ID);
+        }
```
Comment (ChloeSartorelli) [inline src/Services/QueuedJobService.php]:
```suggestion
        $message = 'The following job(s) appear to be broken.';
```
You can use `_t` if you want - the point is it should just be the message part, not the full template.

File: src/Services/EmailService.php [hunk 8]
```diff
@@ -81,12 +69,22 @@ public function createStalledJobReport(string $subject, string $message, int $jo
...
 
+    public function createBrokenJobsReport(string $subject, string $message): ?Email
+    {
+        $email = $this->createReport($subject);
+        if ($email === null) {
+            return null;
+        }
+        return $email
+            ->setData([
+                'Message' => $message,
```
Comment (ChloeSartorelli) [inline src/Services/EmailService.php:81]: Pass in the jobs and site as well, as per suggested change to template.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Services/QueuedJobService.php (hunk 7 above)
Comment (emteknetnz) [reply src/Services/QueuedJobService.php]: Updated - though I've retained most of the message here so that's all translatable

File: src/Services/EmailService.php (hunk 8 above)
Comment (emteknetnz) [reply src/Services/EmailService.php:81]: Updated

File: src/Services/QueuedJobService.php [hunk 9]
```diff
@@ -519,6 +519,21 @@ public function checkJobHealth($queue = null)
             ]
         );
 
+        // Send broken job email notification
+        $subject = _t(__CLASS__ . '.BROKEN_JOBS', 'Broken job(s)');
+        $message = sprintf(_t(
+            __CLASS__ . '.BROKEN_JOBS_MSG',
+            implode(' ', [
+                'The following job(s) appear to be broken.',
+                'Log in to %s to see further details and take any necessary actions.',
```
Comment (ChloeSartorelli) [inline src/Services/QueuedJobService.php]: The other templates have the "log in to etc etc" part in the template. We should aim to have this template be similar to the existing templates.

If you really want to use i18n on it, you should use the template `<%t %>` and update the other templates to also use it. See https://docs.silverstripe.org/en/5/developer_guides/i18n/#usage-in-template-files for docs on how to use i18n in templates.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Services/QueuedJobService.php (hunk 9 above)
Comment (emteknetnz) [reply src/Services/QueuedJobService.php]: I don't feel strongly about it, I've updated it to make it similar to the other emails

**Review by GuySartorelli — APPROVED** [review]
LGTM
